libass/ass_filesystem.c

```diff
@@ -0,0 +1,410 @@
+
+FILE *ass_open_file(const char *filename, FileNameSource hint)
+{
+    return fopen(filename, "rb");
+}
+
+bool ass_open_dir(ASS_Dir *dir, const char *path)
+{
+    dir->handle = NULL;
+    dir->path = NULL;
+    dir->name = NULL;
+
+    size_t len = strlen(path);
+    if (len && path[len - 1] == '/')
+        len--;
+
+    size_t size = NAME_BUF_SIZE + 2;
+    if (!check_add_size(&size, len))
+        return false;
+    dir->path = malloc(size);
+    if (!dir->path)
+        return false;
+    dir->max_path = size;
+    memcpy(dir->path, path, len);
+    dir->path[len] = '/';
+    dir->prefix = len + 1;
+
+    dir->handle = opendir(path);
+    if (dir->handle)
+        return true;
+
+    free(dir->path);
+    dir->path = NULL;
+    return false;
+}
+
+const char *ass_read_dir(ASS_Dir *dir)
+{
+    struct dirent *entry = readdir(dir->handle);
+    return dir->name = entry ? entry->d_name : NULL;
+}
+
+const char *ass_current_file_path(ASS_Dir *dir)
+{
+    size_t size = dir->prefix + 1, len = strlen(dir->name);
+    if (!check_add_size(&size, len) || !alloc_path(dir, size))
+        return NULL;
+    memcpy(dir->path + dir->prefix, dir->name, len + 1);
+    return dir->path;
+}
+
+void ass_close_dir(ASS_Dir *dir)
+{
+    if (dir->handle)
+        closedir(dir->handle);
+    free(dir->path);
+    dir->handle = NULL;
+    dir->path = NULL;
+}
+
+
+#else  // Windows
+
+#include <windows.h>
+#include "ass_directwrite.h"  // for ASS_WINAPI_DESKTOP
+
+
+static const uint8_t wtf8_len_table[256] = {
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  // 0x
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  // 1x
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  // 2x
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  // 3x
+
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  // 4x
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  // 5x
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  // 6x
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  // 7x
+
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  // 8x
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  // 9x
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  // Ax
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  // Bx
+
+    0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,  // Cx
+    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,  // Dx
+    3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,  // Ex
+    4, 4, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  // Fx
+};
+
+static const uint8_t wtf8_len4_range[5][2] = {
+    { 0x90, 0x30 }, { 0x80, 0x40 }, { 0x80, 0x40 }, { 0x80, 0x40 }, { 0x80, 0x10 }
+};
+
+static inline bool check_add_size_wtf8to16(size_t *size, size_t len)
+{
+    return len > SIZE_MAX / sizeof(wchar_t) ? false : check_add_size(size, len * sizeof(wchar_t));
+}
+
+// does not append zero termination implicitly
+// expects preallocated buffer (use check_add_size_wtf8to16() for size)
+static wchar_t *convert_wtf8to16(wchar_t *dst, ASS_StringView src)
```

Comment:
Can you replace all instances of `wchar_t` with Microsoft's `WCHAR` as used in Microsft's headers? Those are different types in midipix, but here all of them are asumed to be 16-bit and meant to be passed to Win32-APIs.

When the widths of `wchar_t` and `WCHAR` differ, there's also a problem with using `WCHAR` with standard `wchar_t`-APIs, in this patchset: `wcslen`, and `L"..."`-literals, but since midipix isn't fully public yet, we concluded in #585 to ignore those two problems for now, but already start using `WCHAR` when Win32-APIs are involved for consistency with MS-headers and to better convey the width-assumption.

Reply:
Done. But midipix is already working in UTF8, so we should use our *nix code path for it anyway.